File: examples/python/apps/miniclaw/orchestrator.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""OrchestratorActor — durable workflow that decomposes tasks and aggregates results.

Dispatches sub-tasks to agents discovered via process groups, then writes
intermediate results to TupleSpace for coordination.
"""

from plexspaces import workflow_actor, state, run_handler, signal_handler, query_handler, init_handler, host
from .helpers import pg_first, fire_audit, ask


@workflow_actor
class OrchestratorActor:
    """Durable workflow: decompose task → delegate to agents → aggregate results."""

    status: str = state(default="idle")
    task_id: str = state(default="")
    progress: int = state(default=0)
    actor_id: str = state(default="")

    @init_handler
    def on_init(self, config: dict) -> None:
        self.actor_id = config.get("actor_id", "")
        host.info(f"OrchestratorActor init actor_id={self.actor_id}")

    @run_handler
    def run(self, payload: dict = None) -> dict:
        payload = payload or {}
        task = payload.get("task", "explain how agents work")
        task_id = payload.get("task_id", f"orch-{host.now_ms()}")

        self.status = "running"
        self.task_id = task_id
        self.progress = 0
        host.info(f"Orchestrator run task_id={task_id} task={task}")

        agent_id, err = pg_first("svc:agent")
        if err or not agent_id:
            self.status = "failed"
            return {"error": "no agents in svc:agent", "task_id": task_id}

        # Decompose: split on " and " for multi-step tasks
        lower = task.lower()
        idx = lower.find(" and ")
        sub_tasks = [task[:idx].strip(), task[idx + 5:].strip()] if idx >= 0 else [task]

        sub_results = []
        for i, sub_task in enumerate(sub_tasks):
            self.progress = (i + 1) * 100 // len(sub_tasks)
            resp = ask(agent_id, "chat", {"message": sub_task, "session_id": f"orch-{task_id}-{i}"}, 15000)
            if not resp:
                self.status = "failed"
                return {"error": "sub-task failed", "task_id": task_id}
            host.ts.write(["orch_result", task_id, i, str(resp.get("response", ""))])
            sub_results.append(resp)

        summaries = [r.get("response", "") for r in sub_results if r.get("response")]
        aggregated = " | ".join(summaries)

        self.status = "completed"
        self.progress = 100
        fire_audit("orchestrator_completed", f"task_id={task_id} subtasks={len(sub_tasks)}")
        return {
            "status": "ok",
            "task_id": task_id,
            "result": aggregated,
            "sub_results": sub_results,
            "sub_tasks": len(sub_tasks),
        }

    @signal_handler("cancel")
    def cancel(self) -> None:
        self.status = "cancelled"
        host.info(f"Orchestrator cancelled task_id={self.task_id}")

    @query_handler("status")
    def query_status(self) -> dict:
        return {"task_id": self.task_id, "status": self.status, "progress": self.progress}

File: examples/go/apps/miniclaw/orchestrator.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// OrchestratorActor — durable workflow that decomposes tasks and delegates to agents.
package main

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/bhatti/PlexSpaces/sdks/go/plexspaces"
)

// OrchestratorActor is a durable workflow that decomposes a task into sub-tasks,
// delegates each to an AgentActor discovered via process group, and aggregates
// results via TupleSpace coordination.
//
// The framework routes workflow_run/workflow_signal/workflow_query to Run/Signal/Query
// automatically. Handle() must not dispatch these itself.
type OrchestratorActor struct {
	plexspaces.BaseActor
	Status   string `json:"status"`
	TaskID   string `json:"task_id"`
	Progress int    `json:"progress"`
}

func NewOrchestratorActor() plexspaces.Actor {
	a := &OrchestratorActor{Status: "idle"}
	a.SetSelf(a)
	return a
}

func newOrchestratorActor() *OrchestratorActor {
	a := &OrchestratorActor{Status: "idle"}
	a.SetSelf(a)
	return a
}

func (o *OrchestratorActor) Init(configJSON string) string {
	var config struct {
		ActorID string `json:"actor_id"`
	}
	if err := json.Unmarshal([]byte(configJSON), &config); err != nil {
		return "ERROR: " + err.Error()
	}
	o.SetRuntimeMetadata(config.ActorID)
	o.Status = "idle"
	host.Info(fmt.Sprintf("OrchestratorActor Init actor_id=%s", config.ActorID))
	return ""
}

// Handle must NOT dispatch workflow_run/signal/query — the framework does that.
func (o *OrchestratorActor) Handle(from, msgType, payloadJSON string) string {
	p := parsePayload(payloadJSON)
	switch msgType {
	case "workflow_run":
		return o.Run(payloadJSON)
	case "workflow_query":
		name := stringVal(p, "name", "status")
		return o.Query(name, payloadJSON)
	case "workflow_signal":
		name := stringVal(p, "name", "")
		o.Signal(name, payloadJSON)
		return marshal(map[string]any{"ok": true})
	}
	return marshal(map[string]any{"error": "use workflow_run / workflow_signal / workflow_query"})
}

func (o *OrchestratorActor) Run(payloadJSON string) string {
	p := parsePayload(payloadJSON)
	task := stringVal(p, "task", "explain how agents work")
	taskID := stringVal(p, "task_id", fmt.Sprintf("orch-%d", host.NowMs()))

	o.Status = "running"
	o.TaskID = taskID
	o.Progress = 0

	host.Info(fmt.Sprintf("OrchestratorActor Run taskID=%s task=%s", taskID, task))

	agentID, err := pgFirst("svc:agent")
	if err != nil {
		o.Status = "failed"
		return marshal(map[string]any{"error": "no agents in svc:agent process group", "task_id": taskID})
	}

	// Decompose task into sub-tasks by splitting on " and ".
	subTasks := []string{}
	if idx := strings.Index(strings.ToLower(task), " and "); idx >= 0 {
		subTasks = append(subTasks, strings.TrimSpace(task[:idx]))
		subTasks = append(subTasks, strings.TrimSpace(task[idx+5:]))
	} else {
		subTasks = []string{task}
	}

	subResults := make([]any, 0, len(subTasks))
	for i, subTask := range subTasks {
		o.Progress = (i + 1) * 100 / len(subTasks)
		resp, err := host.Ask(agentID, "chat", map[string]any{
			"op":         "chat",
			"message":    subTask,
			"session_id": fmt.Sprintf("orch-%s-%d", taskID, i),
		}, 15000)
		if err != nil {
			o.Status = "failed"
			return marshal(map[string]any{"error": "sub-task failed: " + err.Error(), "task_id": taskID})
		}
		resultJSON, _ := json.Marshal(resp)
		_ = host.TS().Write([]any{"orch_result", taskID, i, string(resultJSON)})
		subResults = append(subResults, resp)
	}

	summaries := make([]string, 0, len(subResults))
	for _, r := range subResults {
		if rm, ok := r.(map[string]any); ok {
			if response := stringVal(rm, "response", ""); response != "" {
				summaries = append(summaries, response)
			}
		}
	}
	aggregated := strings.Join(summaries, " | ")

	o.Status = "completed"
	o.Progress = 100

	o.IncrCounter(host, "orchestrator_runs")
	fireAudit("orchestrator_completed", fmt.Sprintf("task_id=%s subtasks=%d", taskID, len(subTasks)))
	return marshal(map[string]any{
		"status":      "ok",
		"task_id":     taskID,
		"result":      aggregated,
		"sub_results": subResults,
		"sub_tasks":   len(subTasks),
	})
}

func (o *OrchestratorActor) Signal(name, payloadJSON string) {
	switch name {
	case "cancel":
		o.Status = "cancelled"
		host.Info(fmt.Sprintf("OrchestratorActor cancelled task_id=%s", o.TaskID))
	}
}

func (o *OrchestratorActor) Query(name, _ string) string {
	if name == "status" {
		return marshal(map[string]any{
			"task_id":  o.TaskID,
			"status":   o.Status,
			"progress": o.Progress,
		})
	}
	return marshal(map[string]any{"error": "unknown_query", "name": name})
}
